### PR TITLE
Bechmarks

### PR DESCRIPTION
**Pull request by LegendreM**
Add several benchmarks:
- initialization: how much time each pipeline takes to initialize
- default_run: how much time our default pipeline takes to analyze different texts
- normalizer:  how much time each normalizer takes to normalize different texts
- tokenizer:  how much time each tokenizer takes to tokenize different texts

### CODE REVIEW

File: benches/default_run.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    // analyze a first time each text to trigger lazy initializations
+    for &(_name, text) in data_set {
+        analyzer.analyze(text);
+    }
+
+    let mut group = c.benchmark_group("default-run");
+
+    for &(name, text) in data_set {
+        group.bench_with_input(BenchmarkId::new("default-run", name), &(&analyzer, black_box(text)), |b, &(a, s)| b.iter(|| run(a, s)));
```
Comment (MarinPostma) [inline benches/default_run.rs:18]: what exactly is the difference between `bench_with_input` and simply capturing `(&analyzer, black_box(text))` in a `bench_function`?

File: benches/default_run.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    group.finish();
+}
+
+fn run(analyzer: &Analyzer<Vec<u8>>, text: &str) {
+
+    let analyzed = analyzer.analyze(text);
+    
+    black_box::<Vec<Token>>(analyzed.tokens().collect());
```
Comment (MarinPostma) [inline benches/default_run.rs:28]: you are benchmarking allocation here too, wouldn't a count be enough to consume the iterator here?

File: benches/normalizer.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    let analyzer = init_analyzer_with_normalizer(LowercaseNormalizer, &stop_words);
+    for &(name, text) in data_set {
+        group.bench_with_input(BenchmarkId::new("LowercaseNormalizer", name), &(&analyzer, black_box(text)), |b, &(a, s)| b.iter(|| run(a, s)));
+    }
+
+    group.finish();
+}
+
+fn run(analyzer: &Analyzer<Vec<u8>>, text: &str) {
+
```
Comment (MarinPostma) [inline benches/normalizer.rs:49]:
```suggestion
```

File: benches/tokenizer.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    let analyzer = init_analyzer_with_tokenizer(Jieba, &stop_words);
+    for &(name, text) in data_set {
+        group.bench_with_input(BenchmarkId::new("Jieba", name), &(&analyzer, black_box(text)), |b, &(a, s)| b.iter(|| run(a, s)));
+    }
+
+    group.finish();
+}
+
+fn run(analyzer: &Analyzer<Vec<u8>>, text: &str) {
+
```
Comment (MarinPostma) [inline benches/tokenizer.rs:48]:
```suggestion
```

**Review by MarinPostma — CHANGES_REQUESTED** [review]

File: benches/default_run.rs (hunk 1 above)
Comment (LegendreM) [reply benches/default_run.rs:18]: what do you mean by "capturing in a `bench_function`"?

File: benches/default_run.rs (hunk 2 above)
Comment (LegendreM) [reply benches/default_run.rs:28]: you are right it should be better! 👍

File: benches/normalizer.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    let analyzed = analyzer.analyze(text);
+    
+    black_box::<Vec<Token>>(analyzed.tokens().collect());
```
[the remark this answers is not in the document]
Comment (LegendreM) [reply benches/normalizer.rs:51]: 👍

File: benches/default_run.rs (hunk 1 above)
Comment (MarinPostma) [reply benches/default_run.rs:18]: I mean you are basically using a closure here: `|b, &(a, s)| b.iter(|| run(a, s)` and passing (a, s) to it, why not directly do this: 
```suggestion
        group.bench_with_function(BenchmarkId::new("default-run", name), |b| b.iter(|| run(&analyzer, black_box(text)));
```

Comment (LegendreM) [reply benches/default_run.rs:18]: don't know, I followed the `tutorial`

Comment (LegendreM) [reply benches/default_run.rs]: Ok the difference is that `bench_with_input use the` black_box()` function on the input before running the benchmark

**Review by MarinPostma — APPROVED** [review]
good job!

**Pull request merged**